### PR TITLE
ISPN-14568 Cache Alias

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-14568

### CODE REVIEW

File: core/src/main/java/org/infinispan/configuration/cache/Configuration.java [hunk 1]
```diff
@@ -19,9 +23,16 @@
 public class Configuration extends ConfigurationElement<Configuration> implements BasicConfiguration {
    public static final AttributeDefinition<String> CONFIGURATION = AttributeDefinition.builder(org.infinispan.configuration.parsing.Attribute.CONFIGURATION, null, String.class).immutable().build();
    public static final AttributeDefinition<Boolean> SIMPLE_CACHE = AttributeDefinition.builder(org.infinispan.configuration.parsing.Attribute.SIMPLE_CACHE, false).immutable().build();
+   public static final AttributeDefinition<List<String>> ALIASES = AttributeDefinition.builder(org.infinispan.configuration.parsing.Attribute.ALIASES, null, (Class<List<String>>) (Class<?>) List.class)
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/configuration/cache/Configuration.java:26]: `@SuppressWarnings("unchecked")`?

File: core/src/main/java/org/infinispan/configuration/ConfigurationManager.java [hunk 2]
```diff
@@ -119,4 +130,55 @@ public ConfigurationBuilderHolder toBuilderHolder() {
...
+
+   private void removeAliases(Collection<String> aliases) {
+      for (String alias : aliases) {
+         this.aliases.remove(alias);
+      }
+   }
+
+   public String selectCache(String cacheName) {
+      String name = aliases.get(cacheName);
+      return name != null ? name : cacheName;
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/configuration/ConfigurationManager.java:161]: `return aliases.getOrDefault(cacheName, cacheName);`

File: core/src/main/java/org/infinispan/configuration/ConfigurationManager.java [hunk 3]
```diff
@@ -89,7 +93,10 @@ private Configuration findConfiguration(String name) {
    }
 
    public Configuration putConfiguration(String cacheName, Configuration configuration) {
+      // This will fail if there are already
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/configuration/ConfigurationManager.java:96]: "...already aliases defined"?

File: core/src/main/java/org/infinispan/util/logging/Log.java [hunk 4]
```diff
@@ -2386,4 +2386,6 @@ CacheConfigurationException storeConfiguredHasBothReadAndWriteOnly(String storeC
    @Message(value = "Security container cannot be enabled/disabled at cache level, since it is configured globally.", id = 974)
    CacheConfigurationException securityCacheTracing();
 
+   @Message(value = "The alias '%s' is already being used for cache '%s'", id = 975)
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/util/logging/Log.java:2389]: "used by cache"?

File: core/src/test/java/org/infinispan/api/CacheAPITest.java [hunk 5]
```diff
@@ -17,18 +17,18 @@
...
 import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.configuration.cache.IsolationLevel;
 import org.infinispan.manager.EmbeddedCacheManager;
 import org.infinispan.test.TestingUtil;
 import org.infinispan.test.fwk.TestCacheManagerFactory;
-import org.infinispan.configuration.cache.IsolationLevel;
 import org.testng.annotations.Test;
 
+import jakarta.transaction.TransactionManager;
+
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/api/CacheAPITest.java:31]: Unrelated change, can be reverted.

File: core/src/test/java/org/infinispan/api/CacheAliasTest.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+   public void testCacheAliases() {
+      cacheManager.getCache("cacheA").put(k(), v());
+      cacheManager.getCache("cacheB").put(k(), v(1));
+      assertEquals(v(), cacheManager.getCache("0").get(k()));
+      assertEquals(v(1), cacheManager.getCache("1").get(k()));
+   }
+
+   public void testAliasConflict() {
+      Exceptions.expectException(CacheConfigurationException.class, "^ISPN000974:.*", () ->
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/api/CacheAliasTest.java:39]: The test is failing, the regex needs to be updated to `^ISPN000975:.*` with the current implementation.

File: core/src/test/java/org/infinispan/api/CacheAliasTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import org.infinispan.test.fwk.TransportFlags;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "api.CacheAliasTest")
+public class CacheAliasTest extends SingleCacheManagerTest {
+
+   @Override
+   protected EmbeddedCacheManager createCacheManager() throws Exception {
+      GlobalConfigurationBuilder globalBuilder = new GlobalConfigurationBuilder().nonClusteredDefault();
+      amendGlobalConfiguration(globalBuilder, new TransportFlags());
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/api/CacheAliasTest.java:24]: Line can be removed.

File: core/src/test/java/org/infinispan/commands/module/TestGlobalConfigurationBuilder.java [hunk 8]
```diff
@@ -47,7 +48,9 @@ public TestGlobalConfigurationBuilder testCacheComponent(String cacheName, Strin
    }
 
    public TestGlobalConfigurationBuilder addCacheStartingCallback(Consumer<ComponentRegistry> callback) {
-      this.attributes.attribute(TestGlobalConfiguration.CACHE_STARTING_CALLBACKS).get().add(callback);
+      List<Consumer<ComponentRegistry>> callbacks = this.attributes.attribute(TestGlobalConfiguration.CACHE_STARTING_CALLBACKS).get();
+      callbacks.add(callback);
+      this.attributes.attribute(TestGlobalConfiguration.CACHE_STARTING_CALLBACKS).set(callbacks);
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/commands/module/TestGlobalConfigurationBuilder.java:53]: Why is this change necessary?

File: server/runtime/src/main/server/server/conf/infinispan-dev-mode.xml [hunk 9]
```diff
@@ -36,11 +36,11 @@
    </jgroups>
 
    <cache-container name="default" statistics="true">
-      <metrics accurate-size="true"/>
       <transport cluster="${infinispan.cluster.name:cluster}" stack="${infinispan.cluster.stack:tcp}" node-name="${infinispan.node.name:}"/>
       <security>
          <authorization/>
       </security>
+      <metrics accurate-size="true"/>
```
Comment (ryanemerson) [inline server/runtime/src/main/server/server/conf/infinispan-dev-mode.xml:43]: Unrelated change, can be reverted.

File: server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java [hunk 10]
```diff
@@ -28,10 +28,10 @@
 import java.util.Calendar;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.LinkedHashMap;
 import java.util.List;
-import java.util.Set;
 import java.util.Map;
-import java.util.LinkedHashMap;
+import java.util.Set;
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java:34]: Unrelated change, can be reverted.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/SELECT.java [hunk 11]
```diff
@@ -24,8 +29,16 @@ public SELECT() {
...
+      String db = new String(arguments.get(0), StandardCharsets.US_ASCII);
+      Cache<byte[], byte[]> cache = handler.respServer().getCacheManager().getCache(db);
+      if (cache != null) {
+         ConnectionMetadata metadata = ConnectionMetadata.getInstance(ctx.channel());
+         handler.setCache(cache.getAdvancedCache().withSubject(metadata.subject()).withMediaType(MediaType.APPLICATION_OCTET_STREAM, MediaType.APPLICATION_OCTET_STREAM));
+         metadata.cache(cache.getName());
+         Consumers.OK_BICONSUMER.accept(null, handler.allocator());
+      } else {
+         ByteBufferUtils.stringToByteBuf("-ERR DB index is out of range\r\n", handler.allocator());
+      }
```
Comment (ryanemerson) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/SELECT.java:41]: My Redis/RESP knowledge is lacking here. I understand that we're adding an alias of `0` by default, but what must be configured in order for an Infinispan user to do `SELECT 1`? This should be documented.

**Review by ryanemerson — CHANGES_REQUESTED** [review]

File: core/src/main/resources/schema/infinispan-config-15.0.xsd [hunk 12]
```diff
@@ -357,6 +357,7 @@
             <xs:documentation>Defines a DIST_* mode cache configuration.</xs:documentation>
           </xs:annotation>
         </xs:element>
+        <xs:any namespace="##other"/>
```
Comment (wburns) [inline core/src/main/resources/schema/infinispan-config-15.0.xsd:360]: Do we expect a custom namespace for a different cache type?

File: server/resp/src/main/java/org/infinispan/server/resp/CacheRespRequestHandler.java [hunk 13]
```diff
@@ -8,8 +8,7 @@ public class CacheRespRequestHandler extends RespRequestHandler {
 
    protected CacheRespRequestHandler(RespServer respServer) {
       super(respServer);
-      setCache(respServer.getCache()
-            .withMediaType(MediaType.APPLICATION_OCTET_STREAM, MediaType.APPLICATION_OCTET_STREAM));
+      setCache(respServer.getCache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, MediaType.APPLICATION_OCTET_STREAM));
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/CacheRespRequestHandler.java:11]: Can be reverted, no reason to change.

File: server/resp/src/main/java/org/infinispan/server/resp/RespServer.java [hunk 14]
```diff
@@ -90,7 +90,7 @@ protected void startInternal() {
             }
             builder.encoding().key().mediaType(RESP_KEY_MEDIA_TYPE);
             builder.encoding().value().mediaType(configuredValueType);
-            builder.statistics().enable();
+            builder.statistics().enable().aliases("0");
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/RespServer.java:93]: Just seeking clarification, but will RESP explicitly ask for cache 0 ? And this should probably outside of this else so it always applied whenever an explicit configuration isn't supplied right?

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java [hunk 15]
```diff
@@ -942,7 +942,7 @@ private CompletionStage<RestResponse> getCacheConfigMutableAttribute(RestRequest
    private CompletionStage<RestResponse> setCacheConfigMutableAttribute(RestRequest request) {
       NettyRestResponse.Builder responseBuilder = invocationHelper.newResponse(request);
       String attributeName = request.getParameter("attribute-name");
-      String attributeValue = request.getParameter("attribute-value");
+      String attributeValue = String.join(" ", request.parameters().get("attribute-value"));
```
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java:945]: This seems like a pretty big change for other things that take parameters as well outside of this...

**Review by wburns — COMMENTED** [review]
Looking good, just some minor concerns

File: core/src/test/java/org/infinispan/api/CacheAPITest.java (hunk 5 above)
Comment (tristantarrant) [reply core/src/test/java/org/infinispan/api/CacheAPITest.java:31]: I have "organize imports" as a save action :)

File: server/resp/src/main/java/org/infinispan/server/resp/RespServer.java (hunk 14 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/RespServer.java:93]: RESP doesn't explicitly ask for it. I'd also prefer not to enforce it in other cases

File: core/src/main/resources/schema/infinispan-config-15.0.xsd (hunk 12 above)
Comment (tristantarrant) [reply core/src/main/resources/schema/infinispan-config-15.0.xsd:360]: I think this is a leftover of the grander plan I had for cache aliases. I'll revert this

File: server/resp/src/main/java/org/infinispan/server/resp/RespServer.java (hunk 14 above)
Comment (wburns) [reply server/resp/src/main/java/org/infinispan/server/resp/RespServer.java:93]: Okay, that is what I figured, I still think we need to set the alias outside the else block so it is applied whenever an explicit configuration isn't supplied though.

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java (hunk 15 above)
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceV2.java:945]: `attribute.fromString()` already supports splitting space-separated multi-value attributes (this is how XML attributes work), so not really a big change.

File: server/resp/src/main/java/org/infinispan/server/resp/RespServer.java (hunk 14 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/RespServer.java:93]: Ok

File: server/resp/src/main/java/org/infinispan/server/resp/CacheRespRequestHandler.java (hunk 13 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/CacheRespRequestHandler.java:11]: ok

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java [hunk 16]
```diff
@@ -229,6 +234,11 @@ public Builder<T> initializer(AttributeInitializer<? extends T> initializer) {
          return this;
       }
 
+      public Builder<T> matcher(AttributeMatcher<T> matcher) {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:237]: At the moment the need is only theoretical.

File: client/rest-client/src/main/java/org/infinispan/client/rest/impl/jdk/RestCacheClientJDK.java [hunk 17]
```diff
@@ -476,8 +476,15 @@ public CompletionStage<RestResponse> updateWithConfiguration(RestEntity configur
    }
 
    @Override
-   public CompletionStage<RestResponse> updateConfigurationAttribute(String attribute, String value) {
-      return client.post(path + "?action=set-mutable-attribute&attribute-name=" + attribute + "&attribute-value=" + sanitize(value));
+   public CompletionStage<RestResponse> updateConfigurationAttribute(String attribute, String... value) {
+      StringBuilder sb = new StringBuilder(path);
+      sb.append("?action=set-mutable-attribute&attribute-name=");
+      sb.append(attribute);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply client/rest-client/src/main/java/org/infinispan/client/rest/impl/jdk/RestCacheClientJDK.java]: Done

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/SELECT.java (hunk 11 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/connection/SELECT.java:41]: I've added a chapter to the redis connector docs

File: server/runtime/src/main/server/server/conf/infinispan-dev-mode.xml (hunk 9 above)
Comment (tristantarrant) [reply server/runtime/src/main/server/server/conf/infinispan-dev-mode.xml]: Done

File: core/src/test/java/org/infinispan/api/CacheAliasTest.java (hunk 7 above)
Comment (tristantarrant) [reply core/src/test/java/org/infinispan/api/CacheAliasTest.java]: Done

File: core/src/test/java/org/infinispan/commands/module/TestGlobalConfigurationBuilder.java (hunk 8 above)
Comment (tristantarrant) [reply core/src/test/java/org/infinispan/commands/module/TestGlobalConfigurationBuilder.java:53]: It's because it marks the attribute as modified

File: core/src/main/java/org/infinispan/util/logging/Log.java (hunk 4 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/util/logging/Log.java]: Done

File: core/src/main/java/org/infinispan/configuration/ConfigurationManager.java (hunk 3 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/configuration/ConfigurationManager.java]: Done

File: core/src/main/java/org/infinispan/configuration/ConfigurationManager.java (hunk 2 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/configuration/ConfigurationManager.java]: Done

File: core/src/main/java/org/infinispan/configuration/cache/Configuration.java [hunk 18]
```diff
@@ -303,4 +319,13 @@ public String toStringConfiguration(String name, MediaType mediaType, boolean cl
       }
       return sw.toString();
    }
+
+   public static final class AliasMatcher implements AttributeMatcher<List<String>> {
+      static final AliasMatcher INSTANCE = new AliasMatcher();
+
+      @Override
+      public boolean matches(Attribute<List<String>> o1, Attribute<List<String>> o2) {
+         return true;
```
Comment (fax4ever) [inline core/src/main/java/org/infinispan/configuration/cache/Configuration.java:328]: what's its purpose if it returns always true?

Comment (tristantarrant) [reply core/src/main/java/org/infinispan/configuration/cache/Configuration.java]: I have reworked this a little to provide a better explanation why we're doing this

**Comment by tristantarrant** [comment]
Failures are unrelated

**Comment by tristantarrant** [comment]
It's green!

**Comment by ryanemerson** [comment]
Thanks @tristantarrant
